### ISSUE TITLE
Optimize starup time: Tomcat/Web

### ISSUE THREAD

**Issue by lldata** [spring-projects/spring-boot#8216]
Starting a Spring Boot application with Tomcat/Web is too slow

Generating Spring Boot applications from http://start.spring.io/ and measuring startup on my laptop:

Empty Spring Boot app: 1.5s
Spring Boot with Web: 5.9s
Spring Boot with Web, Actuator: 6.5s

For reference a plain, empty Tomcat 8.5 server: 1.5s

My production app, which also includes cloud components Eureka, Hysterix and Sleuth. Startup time grows to about 30 seconds. This appears to be mainly due to AspectJ and JAR file parsing. I would be happy to give up AutoConfiguration and AspectJ. A little more reluctant to give up ComponentScan. But if a startup time of a production app could be reduced to a few seconds it would be worth it.

The main issue here is just the baseline. Even cutting away most of the magic, startup is still slow.

Why is Spring Boot with Tomcat so slow to start?

IMHO a Spring Boot app with Tomcat should have no noticable overhead to a Tomcat Server.

**Comment by wilkinsona** [comment]
> IMHO a Spring Boot app with Tomcat should have no noticable overhead to a Tomcat Server

That would be nice, but it's not a realistic goal. When you start a Spring Boot web application it's starting an embedded Tomcat server and an application. In other words, it's far from being a like-for-like comparison.

>  This appears to be mainly due to AspectJ and JAR file parsing

It sounds like you've done some analysis and perhaps profiling. Why not share the results of that? Performance issues are impossible to address without lots of data.

We've done quite a bit of profiling recently. The JMH benchmarks and some results on assorted hardware are [available here](https://github.com/dsyer/spring-boot-startup-bench). As a result of this, we did some work in 1.5 to optimise startup time. What version of Spring Boot are you using?

> Why is Spring Boot with Tomcat so slow to start?

It's impossible to say why it's slow for you. For comparison, I can start Spring Boot's Freemarker sample on my laptop (a 3 year old MackBook Pro) in ~2 seconds when run as a fat jar. We have seen significant slow down in the past caused by a virus scanner.  Perhaps you're running one?

**Comment by lldata** [comment]
> When you start a Spring Boot web application it's starting an embedded Tomcat server and an application.

Ok. Then starting Spring Boot with an embedded tomcat should add up to at most the same time as starting a plain Spring Boot app + starting an empty Tomcat

> It sounds like you've done some analysis and perhaps profiling. Why not share the results of that? 

I am trying to make the point that the performance should be better. Even if I don't have the skills to make it so.

I did some profiling on my own app. I find profiling startup times a very hard discipline.
To get something more comparable I profiled a spring boot + web demo application with jprofiler on my mac.

I've attached a call tree, which shows the hot spots from the main thread. Seems like a lot of time is spent waiting for the hostname and the PID. Could that not be optimized?

> We have seen significant slow down in the past caused by a virus scanner. Perhaps you're running one?

Spring Boot plus Web is taking about 12 seconds to start on my two year old top shelf MBP. No virus scanner.

Numbers in my first post are from my windows laptop at work. I am not 100% about how the virus scanner works there. But numbers are lower than the macbook ... It is also newer.

**Comment by lldata** [comment]
Main thread results from JProfiler
[Call_Tree_main.html.zip](https://github.com/spring-projects/spring-boot/files/758541/Call_Tree_main.html.zip)

**Comment by snicoll** [comment]
There is a known issue with network settings that has nothing to do with neither Tomcat nor Spring Boot. Perhaps you [should check if you're affected by the issue](http://stackoverflow.com/questions/39636792/jvm-takes-a-long-time-to-resolve-ip-address-for-localhost)...

**Comment by lldata** [comment]
Thank you @snicoll - I had the described problem on my mac. I was also wondering why the results didn't fit what I see at work.

But going back to my original problem:
Measurements from my MBP (without any JVM flags):
- plain tomcat 8.5.11: 0.6s
- plain spring boot: 1.0s
- spring boot + web: 3.0s

I still think it is reasonable to expect that Spring Boot + Tomcat should be no slower than 1.6s

**Comment by philwebb** [comment]
> I still think it is reasonable to expect that Spring Boot + Tomcat should be no slower than 1.6s

We've spent quite a bit of time looking at how we can improve startup times and a typical Spring Boot 1.5 Tomcat application starts for me in about 1.5 seconds.

From the limited information you've given it's really impossible to see what you're comparing. Does your plain tomcat 8.5.11 application have any WARs deployed? Does it actually do anything? What does the plain Spring Boot application do? What technology does your "spring boot + web" application use? What version of Spring Boot are you using?

I agree with you that a Spring Boot application shouldn't be significantly slower that comparable WAR deployment, and if you've found that's the case we really need something that we can measure to work out why.

Do you have sample applications that you can share (i.e. war/boot/boot+web) that show the speed difference and will give us something that we can profile?

**Comment by lldata** [comment]
I got tomcat from here:
http://mirrors.rackhosting.com/apache/tomcat/tomcat-8/v8.5.11/bin/apache-tomcat-8.5.11.zip 

[catalina.2017-02-07.txt](https://github.com/spring-projects/spring-boot/files/758756/catalina.2017-02-07.txt)

I does quite a lot. It deploys a root app, examples and a manager application.

For Spring Boot I created the apps here: http://start.spring.io/
"plain" is when you just hit download, without dependencies.
"web" is when you type web in the dependencies text field.

**Comment by wilkinsona** [comment]
> I still think it is reasonable to expect that Spring Boot + Tomcat should be no slower than 1.6s
> "web" is when you type web in the dependencies text field.

That's not a like-for-like comparison then. The web dependency is `spring-boot-starter-web` which brings in Spring MVC. As a result, Boot creates a DispatcherServlet, a number of Filters, a Jackson ObjectMapper, a Hibernate Validator, etc. None of that happens in a plain application. If you don't want to use Spring MVC, then using `spring-boot-starter-tomcat` plus a dependency on `org.springframework:spring-web` would be a better basis for your app and would lower the basic startup time. Similarly if you don't want to use Jackson or Hibernate Validator you could use `spring-boot-starter-web` and exclude them, or use `spring-boot-starter-tomcat` and just add the bits that you do need.

**Comment by lldata** [comment]
Ok. So take this a user feedback/rant, rather than a bug report:

I think that the unasked for ObjectMapper and Validator is just the beginning of the story of helpful stuff, that slows me down. So while I might need those at some point, more helpful stuff is just added with each dependency.

That is more a architecture problem, than something we can triage in a profiler.

This whole story started when I got a report from our ops guys that starting one of our spring boot based apps hogs two cpu cores at almost 100% for more than 30 seconds.

I tried profiling that and discovered by divide and conquer that getting rid of aspectj helped a great deal. I don't use aspectj - it was added in through dependencies. But all in all it is pretty hard to optimize an app that loads more than 10.000 classes to do some pretty simple JMS gymnastics.

So my feedback is that while Spring Boot looks cool on powerpoints, in production it guides you to include too much stuff that you don't really need - and scaling it down is really hard work.

But I couldn't ask you to help profile my production app, that is why I opened this much simplified issue. 

I'll go make sure the virus scanner isn't my problem, upgrade to Spring Boot 1.5.1, look forward to Spring 5.0, figure out how to configure spring cloud without aspectj and try everything else your suggest.

But as a longer term strategy I'll be working towards getting rid of AutoConfiguration to see if that leads to apps that are more manageable.
